In [19]:
import numpy as np
import geopandas as gpd

Data obtained from [here](https://github.com/CityOfNewYork/nyc-geo-metadata/blob/master/Metadata/Metadata_BuildingFootprints.md).

In [2]:
df = gpd.read_file("../data/Building_20221022.zip")

In [3]:
df

NAME      BIN  CNSTRCT_YR  LSTMODDATE   LSTSTATYPE  DOITT_ID  \
0        None  3170958      1925.0  2017-08-22  Constructed     96807   
1        None  5028452      1965.0  2017-08-22  Constructed    326368   
2        None  5078368      1970.0  2017-08-22  Constructed    746627   
3        None  3245111      1928.0  2017-08-22  Constructed    786626   
4        None  4161096      1950.0  2017-08-22  Constructed    746409   
...       ...      ...         ...         ...          ...       ...   
1084652  None  4624077      2021.0  2022-10-21  Constructed   1298417   
1084653  None  1091592      2022.0  2022-10-21  Constructed   1298418   
1084654  None  4620163         NaN  2022-10-21  Constructed   1291595   
1084655  None  3426580      2020.0  2022-10-21  Constructed   1292192   
1084656  None  3427205      2020.0  2022-10-21  Constructed   1294189   

         HEIGHTROOF  FEAT_CODE  GROUNDELEV  SHAPE_AREA  SHAPE_LEN    BASE_BBL  \
0         29.749853     2100.0        40.0         0.0        0.0  3065220021   
1         22.630000     2100.0        39.0         0.0        0.0  5012640036   
2         35.760000     2100.0        51.0         0.0        0.0  5060190091   
3         37.500000     2100.0         6.0         0.0        0.0  3086910048   
4         18.015113     2100.0        93.0         0.0        0.0  4075020005   
...             ...        ...         ...         ...        ...         ...   
1084652   12.000000     5110.0        49.0         0.0        0.0  4046640007   
1084653  420.000000     2100.0        50.0         0.0        0.0  1010000055   
1084654   10.000000     5110.0        90.0         0.0        0.0  4066320001   
1084655   30.000000     2100.0        23.0         0.0        0.0  3064060115   
1084656   70.000000     2100.0        62.0         0.0        0.0  3014740007   

         MPLUTO_BBL  GEOMSOURCE  \
0        3065220021  Photogramm   
1        5012640036  Photogramm   
2        5060190091  Photogramm   
3        3086910048  Photogramm   
4        4075020005  Photogramm   
...             ...         ...   
1084652  4046640007  Other (Man   
1084653  1010000055  Other (Man   
1084654   406632000  Other (Man   
1084655  1064060115  Other (Man   
1084656   301474000  Other (Man   

                                                  geometry  
0        POLYGON ((993508.772 167345.645, 993452.492 16...  
1        POLYGON ((937651.670 172257.299, 937650.883 17...  
2        POLYGON ((930033.534 141941.580, 930007.810 14...  
3        POLYGON ((995046.423 149655.549, 995041.868 14...  
4        POLYGON ((1052344.058 214771.630, 1052356.451 ...  
...                                                    ...  
1084652  POLYGON ((1035196.932 225413.652, 1035216.873 ...  
1084653  POLYGON ((988855.286 215835.382, 988823.961 21...  
1084654  POLYGON ((1034163.213 201090.832, 1034194.246 ...  
1084655  POLYGON ((983488.859 159062.737, 983449.084 15...  
1084656  POLYGON ((1006583.822 183203.196, 1006620.290 ...  

[1084657 rows x 15 columns]

We are mostly interested in highrises, so let's just focus on finding a representative building height that would require a water tower in Manhattan.

So let's find the heights of all of the buildings in Manhattan that are over 6 stories. A 6 story building is about 60 to 75 feet tall—let's just say 60 feet.

We know that a BIN that start with 1 indicates a building that is in Manhattan. 

In [4]:
df[df.HEIGHTROOF > 60]

NAME      BIN  CNSTRCT_YR  LSTMODDATE      LSTSTATYPE  DOITT_ID  \
14       None  3099098      1928.0  2017-08-22     Constructed    111394   
62       None  1018457      1935.0  2017-08-22     Constructed    584238   
65       None  3325583      1958.0  2017-08-22     Constructed    393051   
73       None  1039988      1931.0  2017-08-22     Constructed     66240   
98       None  1003956      1900.0  2017-08-22     Constructed    512541   
...       ...      ...         ...         ...             ...       ...   
1084640  None  1019620      2021.0  2022-10-21     Constructed   1295906   
1084646  None  1090823      2021.0  2022-10-21     Constructed   1296162   
1084651  None  1089694      2022.0  2022-10-21  Initialization   1297393   
1084653  None  1091592      2022.0  2022-10-21     Constructed   1298418   
1084656  None  3427205      2020.0  2022-10-21     Constructed   1294189   

         HEIGHTROOF  FEAT_CODE  GROUNDELEV  SHAPE_AREA  SHAPE_LEN    BASE_BBL  \
14        67.910000     2100.0        37.0         0.0        0.0  3045880001   
62       155.490000     2100.0        37.0         0.0        0.0  1008870001   
65        65.160000     2100.0        12.0         0.0        0.0  3017190001   
73       167.214468     2100.0        52.0         0.0        0.0  1013490035   
98        73.130000     2100.0        39.0         0.0        0.0  1003060013   
...             ...        ...         ...         ...        ...         ...   
1084640  145.000000     2100.0        19.0         0.0        0.0  1009010034   
1084646  520.000000     2100.0        53.0         0.0        0.0  1010000011   
1084651  902.000000     2100.0        33.0         0.0        0.0  1007290051   
1084653  420.000000     2100.0        50.0         0.0        0.0  1010000055   
1084656   70.000000     2100.0        62.0         0.0        0.0  3014740007   

         MPLUTO_BBL  GEOMSOURCE  \
14       3045880001  Photogramm   
62       1008870001  Photogramm   
65       3017190001  Photogramm   
73       1013490035  Photogramm   
98       1003060013  Photogramm   
...             ...         ...   
1084640  1009010034  Other (Man   
1084646  1010007501  Other (Man   
1084651  1007297505  Other (Man   
1084653  1010000055  Other (Man   
1084656   301474000  Other (Man   

                                                  geometry  
14       POLYGON ((1002573.407 180626.949, 1002575.489 ...  
62       POLYGON ((989136.037 210787.392, 989181.569 21...  
65       POLYGON ((998028.218 194162.389, 998033.118 19...  
73       POLYGON ((994153.544 215658.782, 994165.738 21...  
98       POLYGON ((986180.447 200723.052, 986187.771 20...  
...                                                    ...  
1084640  POLYGON ((989183.055 207659.658, 989222.752 20...  
1084646  POLYGON ((988823.961 215852.545, 988880.777 21...  
1084651  POLYGON ((984663.308 213245.306, 984660.244 21...  
1084653  POLYGON ((988855.286 215835.382, 988823.961 21...  
1084656  POLYGON ((1006583.822 183203.196, 1006620.290 ...  

[35536 rows x 15 columns]

In [5]:
mnh_buildings_with_towers_df = df.query('BIN.astype("str").str.startswith("1") & HEIGHTROOF > 60')

In [6]:
mnh_buildings_with_towers_df

NAME      BIN  CNSTRCT_YR  LSTMODDATE      LSTSTATYPE  DOITT_ID  \
62       None  1018457      1935.0  2017-08-22     Constructed    584238   
73       None  1039988      1931.0  2017-08-22     Constructed     66240   
98       None  1003956      1900.0  2017-08-22     Constructed    512541   
153      None  1055733      1905.0  2017-08-22     Constructed    748269   
241      None  1031399      1927.0  2017-08-22     Constructed    267454   
...       ...      ...         ...         ...             ...       ...   
1084636  None  1087083      2000.0  2022-10-21     Constructed   1295940   
1084640  None  1019620      2021.0  2022-10-21     Constructed   1295906   
1084646  None  1090823      2021.0  2022-10-21     Constructed   1296162   
1084651  None  1089694      2022.0  2022-10-21  Initialization   1297393   
1084653  None  1091592      2022.0  2022-10-21     Constructed   1298418   

         HEIGHTROOF  FEAT_CODE  GROUNDELEV  SHAPE_AREA  SHAPE_LEN    BASE_BBL  \
62       155.490000     2100.0        37.0         0.0        0.0  1008870001   
73       167.214468     2100.0        52.0         0.0        0.0  1013490035   
98        73.130000     2100.0        39.0         0.0        0.0  1003060013   
153       70.493402     2100.0        68.0         0.0        0.0  1018440058   
241      151.150000     2100.0        97.0         0.0        0.0  1012000005   
...             ...        ...         ...         ...        ...         ...   
1084636  260.000000     2100.0        44.0         0.0        0.0  1005659021   
1084640  145.000000     2100.0        19.0         0.0        0.0  1009010034   
1084646  520.000000     2100.0        53.0         0.0        0.0  1010000011   
1084651  902.000000     2100.0        33.0         0.0        0.0  1007290051   
1084653  420.000000     2100.0        50.0         0.0        0.0  1010000055   

         MPLUTO_BBL  GEOMSOURCE  \
62       1008870001  Photogramm   
73       1013490035  Photogramm   
98       1003060013  Photogramm   
153      1018440058  Photogramm   
241      1012000005  Photogramm   
...             ...         ...   
1084636  1005659021  Other (Man   
1084640  1009010034  Other (Man   
1084646  1010007501  Other (Man   
1084651  1007297505  Other (Man   
1084653  1010000055  Other (Man   

                                                  geometry  
62       POLYGON ((989136.037 210787.392, 989181.569 21...  
73       POLYGON ((994153.544 215658.782, 994165.738 21...  
98       POLYGON ((986180.447 200723.052, 986187.771 20...  
153      POLYGON ((995102.929 231032.369, 995093.401 23...  
241      POLYGON ((992178.988 225786.552, 992166.821 22...  
...                                                    ...  
1084636  POLYGON ((986937.347 206729.979, 987023.562 20...  
1084640  POLYGON ((989183.055 207659.658, 989222.752 20...  
1084646  POLYGON ((988823.961 215852.545, 988880.777 21...  
1084651  POLYGON ((984663.308 213245.306, 984660.244 21...  
1084653  POLYGON ((988855.286 215835.382, 988823.961 21...  

[19776 rows x 15 columns]

In [12]:
mean_building_height = mnh_buildings_with_towers_df.HEIGHTROOF.mean()
mean_building_height

114.0137004285715

In [13]:
mean_building_area = mnh_buildings_with_towers_df.geometry.area.mean()
mean_building_area

7374.386142360955

In [14]:
BUILDING_STORY_HEIGHT_FT = 10
mean_building_stories = mean_building_height / 10
mean_building_stories

11.401370042857149

In [16]:
usable_building_stories = mean_building_stories - 1
mean_livable_sq_ft_per_building = usable_building_stories * mean_building_area
mean_livable_sq_ft_per_building

76703.71910561413

In [17]:
SQ_FT_PER_PERSON = 300

mean_people_per_building = mean_livable_sq_ft_per_building / SQ_FT_PER_PERSON
mean_people_per_building

255.67906368538044

In [25]:
PER_PERSON_MEAN_WATER_CONSUMPTION_LITER_PER_DAY = 420.18

mean_building_water_consumption_per_day = PER_PERSON_MEAN_WATER_CONSUMPTION_LITER_PER_DAY * mean_people_per_building
mean_building_water_consumption_per_day

107431.22897932316

In [26]:
# We compute the tank outflow per minute as a result of the constant draw from usage.
tank_outflow_l_per_minute = mean_building_water_consumption_per_day / 24 / 60
tank_outflow_l_per_minute

74.60502012452997

In [20]:
MEAN_WATER_TANK_HEIGHT_M = 3.6576
MEAN_WATER_TANK_DIAMETER_M = 3.6576
MEAN_WATER_TANK_RADIUS_M = MEAN_WATER_TANK_DIAMETER_M / 2.0
L_PER_CUBIC_M = 1_000

mean_water_tank_volume_l = np.pi * MEAN_WATER_TANK_RADIUS_M ** 2 * MEAN_WATER_TANK_HEIGHT_M * L_PER_CUBIC_M
mean_water_tank_volume_l

38430.71880174275

In [21]:
# TODO: We can't actually let the tank empty, both to maintain water pressure
# and to adhere to fire codes. So in reality the tank needs to be refilled more
# often than this. Should estimate that somehow.

n_tank_cycles_per_day = mean_building_water_consumption / mean_water_tank_volume_l
n_tank_cycles_per_day

2.795451980316626

In [22]:
hours_per_tank_cycle = 24 / n_tank_cycles_per_day
hours_per_tank_cycle

8.58537373168601

In [27]:
# The amount of time it takes to completely refill the tank from empty. Sources
# indicate that this takes from 2–3 hours.
TANK_FILLING_TIME_MINUTES = 180.0
tank_filling_rate_l_per_minute = mean_water_tank_volume_l / TANK_FILLING_TIME_MINUTES
tank_filling_rate_l_per_minute

213.50399334301525

In [29]:
excess_pump_capacity_l_per_minute = tank_filling_rate_l_per_minute - tank_outflow_l_per_minute
excess_pump_capacity_l_per_minute

138.89897321848528

In [30]:
excess_pump_capacity = excess_pump_capacity_l_per_minute / tank_filling_rate_l_per_minute
excess_pump_capacity

0.6505685024604217

In [33]:
tank_filling_time_with_outflow_hours = TANK_FILLING_TIME_MINUTES * (2 - excess_pump_capacity) / 60
tank_filling_time_with_outflow_hours

4.0482944926187345

In [34]:
total_filling_discharge_period_hours = tank_filling_time_with_outflow_hours + hours_per_tank_cycle
total_filling_discharge_period_hours

12.633668224304746

Ok, so we now know that, for the mean building in Manhattan that requires a
water tank, it has a pumping capacity that exceeds the mean outflow as a result
of water usage by about 65%. That means we can easily outpace the demand and
increase the water level of the tank even as the building is consuming its
average amount of water, though it will slow the ability of the tank to refill.
Our goal is to use the pump as little as possible, so we want to essentially
only run the pump full blast when power is cheapest and then otherwise keep it
completely off.

According to the above, we'd basically have to spend 4 hours filling the tank,
and then we'd get about 8.6 hours for the tank to draw down before we have to
fill it again, just off of normal water usage. This gives us some leeway in our
ability to draw additional water and in our ability to time the pumping and
additional discharge.

Let's think about how increasing water demand would affect this capacity.

Essentially what I'm implying above is that we can afford to discharge about 138
l/m more of water in the mean building before the water tank will on average be
unable to refill faster than it is draining. At 138 l/m of average discharge,
the pump would be required to run constantly to maintain the water level, which
would mean that there would be no benefit to this process as there would be no
ability to advantageously time the pumping or discharging to flatten loads. We'd
just be wasting a bunch of energy.

Meanwhile, if we run the pump only exactly as much as needed to accommodate the
usage of the humans in the building, we of course hit maximal efficiency in that
we're only discharging (and pumping) the water that we need. We can't control
when people are discharging the water, but we can control when we pump and when
we choose to add additional discharge. So the key is to find where we have
leeway in that curve to forecase user demand patterns and to choose when we pump
and when we additionally dischage so as to flatten the demand curve as much as
possible while minimized water waste and ensuring the residents have enough
water.